### DISCUSSION TITLE
Android Studio for iOS Dev - How to launch & Sign a Bundle

### DISCUSSION THREAD

**Discussion by davidakoontz** [skiptools/skip#236] [Q&A] [answered]
I'm struggling with getting a simple HelloSkipper App deployed to the Play Store.

For the last 2 weeks I've been attempting to configure Fastlane - assuming that it would then build/sign/deploy but that has NOT worked out... never got close to success.  So I've punted Fastlane and now I'm back to reading Google's Docs and attempting to get Android Studio to do the magic.  It's not easy either.

Does anyone on this board have Play Store / Android Studio experience to offer advice?  I'm needing help.

Here's what I'm learning... experimenting with:

Skip with create the old APK file - but Play Store wants the NEWER AAB "bundle" file format.

> skip export  --->> get's me a APK file - hidden (for no real reason) in  ProjectFolder/.build folder
❯  ls -lah .build/Android/app/outputs/apk/*/*.apk
-rw-r--r--  1 david  staff    66M Sep 13 10:35 .build/Android/app/outputs/apk/debug/app-debug.apk
-rw-r--r--  1 david  staff    14M Sep 13 10:36 .build/Android/app/outputs/apk/release/app-release-unsigned.apk

If you ignorantly (like me) upload this to Play Store - you will get some RED text ...  so don't try it.... (but I did).

In the case, when you try to upload your APK to the Google Play Console and you are receiving an error message saying, "Upload a valid app bundle." This means that the platform you are trying to upload the app to is expecting an app bundle instead of an APK. To resolve this issue, you need to use the Android Studio's "Build Bundle(s)/APK(s)" menu option or use command line tools such as bundletool to create an app bundle from their existing app.

Now you need to "bundle" the APK...  Skip Docs (if you can find it... say...)
then Use gradle to convert APK to Bundle AAB file in  .build/Android/app/outputs/bundle/release

❯ gradle -p Android/ bundle
[1/1] Compiling plugin Create SkipLink
[2/2] Compiling plugin skipstone
.... all kinds of stuff ... removed for brevity (oh this is going to get longer)...

It will not tell you where your bundle is... again in the HIDDEN folder for no reason under the SUN should a build tool use a hidden folder!

❯ ls -al .build/Android/app/outputs/bundle/release
total 19296
drwxr-xr-x  3 david  staff       96 Sep 13 10:36 .
drwxr-xr-x  4 david  staff      128 Sep 13 10:36 ..
-rw-r--r--  1 david  staff  9551518 Sep 13 10:36 app-release.aab

You - like I may try UPLOADING the new bundle (AAB file) to the Play Store...  again it will FAIL ... now it's about SIGNING it.

I've attempted to setup  Google Play Store... automatic signing ... some thing they recommended... I've NO IDEA if this is set up or NOT... don't get much feedback & now I can't even find the Doc page to name it properly.

Anyway ... to get to the point really fast.  Uploading the bundle gives the error that it is NOT Signed.

So now I'm trying to figure out how to sign / and what to sign with...  All this magic resides in the Android Studio...

In Android Studio - I don't know how to OPEN the project.... internet said that you can NOT click on a file and open the project in A-S.  That's awesome right!  So the basic A-S window has an OPEN ... I find the Android fold and click there... then on the XML Manifest file some where... that appears to work (hint:  IT DOES NOT REALY WORK).

The internet said to goto A-S (that's Android Studio) **Build > Generate Signed App Bundle**  But guess what - it's not on my menu!
Now the internet seems to think that I don't have the "correct" project open...  so down this rabbit trail...
Some how opening the project the correct way... I cannot remember the magic...
And the **Generate Signed App Bundle** appears on the Build menu.

Well now we are off to the races!

... more to come...

Wouldn't it be NICE if there was a Guide to Deploy my skip generated bundle to the Play Store... covering all this... you know an EASY 1, 2, 3.... guide.

**Comment by davidakoontz** [comment, marked as answer]
This whole process seems to be designed back to front… it’s a 2 step process on the Google Doc. 
 But now I’m at least 5 prerequisite into the first STEP.


After getting the PROPER project open in Android-Studio the Build > Generate
<img width="835" alt="Screenshot 2024-09-13 at 1 20 32 PM" src="https://github.com/user-attachments/assets/5d9a71af-5ddf-481f-98b6-78e5f2d9cc49">

menu item will display.  Now you just need a KeyStore!  Where the heck do I get one of those?
I wonder if a *DER file is good for that?

Let’s try it…
<img width="679" alt="People also ask" src="https://github.com/user-attachments/assets/b8787b70-f3fb-4ff6-8aec-b2f0688588ae">
Nope! does not like the *.der file.

A video I watched on how to SIGN an App... suggested "Create NEW"... seems like the way to go at this point of frustration.

<img width="542" alt="New Key Store" src="https://github.com/user-attachments/assets/89eafb0a-51f1-4e62-b67c-ca058d19fa3b">

Create a NEW key store (I guess somewhere in your project would make sense)... fill in a lot of passwords (I just made up new ones) and a bit more info...

Gradle runs for a whiles….
Success!!  Where’s the signed file???  Search for *.aab...

❯ ls Android/app/release
app-release.aab

UPloaded to Play Store  and got download link:

https://play.google.com/apps/internaltest/4700xxxxxREDACTEDxxxxx8

It’s in the Test Play Store!!!


YEAH!  Now see if I can make changes and do it again!

**Comment by davidakoontz** [comment]
Google Play Store help pages:  **Use Play App Signing** 

starts on page: https://github.com/orgs/skiptools/discussions/236

Step 1: Create an upload key
   1. Following these [instructions](https://developer.android.com/studio/publish/app-signing), create an upload key.
   2. Sign your app bundle with the upload key.

Step 2: Prepare your release
   ( 1 ... 4 )
Step 3: Register your app signing key with API providers
   ( 1 & 2 )

It all looks easy enough... it's NOT.  I wish I could have gotten Fastlane to work.  2 weeks vs 1 day

**Reply by marcprux** [reply]
Glad to see that you have made progress on this! Signing is the biggest hassle of preparing a release – both for Android and iOS – and there are dozens of different approaches that one can take. Thanks for taking the time to write up the steps and share the screenshots.

**Reply by davidakoontz** [reply]
I’ve finally deployed to Play store twice now. 
I must be The Expert!
